timApp/plugin/plugin.py

```diff
@@ -722,6 +729,56 @@ def expand_macros_for_plugin(par: DocParagraph, macros, env: TimSandboxedEnviron
     return yaml_str
 
 
+def expand_macros_for_plugin_attrs(
+    par: DocParagraph, macros, env: TimSandboxedEnvironment
+) -> None:
+    """
+    Expands macros for plugin paragraph/block attributes. Sets attribute values directly.
+    In order to set paragraph classes correctly based on certain plugin block attributes
+    (such as lecture questions), this function needs to run before DocParagraph.prepare(..)
+    (see also function pluginify in timApp/plugin/pluginControl.py).
+
+    :param par: The plugin paragraph.
+    :param macros: Dict of macros
+    :param env: macro environment
+    """
+    plugin_type = par.get_attr("plugin")
+    if plugin_type not in ALLOWED_MACRO_PLUGIN_TYPES or par.get_nomacros():
+        return
+    par_attrs = par.get_attrs()
+    expandable_attrs = ""
+    for p_attr in [*par_attrs]:
+        if p_attr in ALLOWED_MACRO_PLUGIN_ATTRS:
+            expandable_attrs += f"{p_attr}: {par.get_attr(p_attr)}, "
```

Comment:
Mielestäni hyvä ajatus ja kätevä optimointi. Mietin kuitenkin, että tämä taitaa aiheuttaa ei-halutun sivuvaikutuksen. Esimerkiksi jos antaa attribuutin arvoksi

```
# {question=", b: 1,  "}
```

niin plugniin tulee uusi attribuutti b:

![image](https://user-images.githubusercontent.com/5202606/225876127-bc610756-0f66-4653-96b9-20fb0bb08940.png)

Tällä hetkellä se ei ole mikään ongelma, mutta jos joskus satutaan käyttämään muitakin attribuutteja.

Olisiko kuinka hidasta tehdä niin, että kutsutaan tuo `expand_macros` jokaisen attribuutin kohdalla erikseen? Eli mallia

```py
for p_attr in par_attrs:
        if p_attr in ALLOWED_MACRO_PLUGIN_ATTRS:
            expanded_val = expand_macros(par.get_attr(p_attr), ....)
            par.set_attr(p_attr, expanded_val)
```

Tai vastaavasti siis.

Reply:
Tuohan tulee nykyisinkin tuotannossa, ainakin kun testasin niin pluginille voi antaa mielivaltaisen attribuutin. Eli tuo ei ole uusi käytös (paitsi että nyt osataan jollakin tasolla käsitellä tunnettujen attribuuttien virheelliset arvot)? 
Tässähän ei ole tarpeen käsitellä mitään muita kuin nuo erikseen sallitut makrotetut attribuutit -- lähinnä sen takia että jotkin pluginit (ainakin qst/luentokysymys) tarvitsevat tietyt css-luokat, mitkä päätetään näiden attribuutin arvon perusteella.
Sitten tuonne `ALLOWED_MACRO_PLUGIN_TYPES` ja `ALLOWED_MACRO_PLUGIN_ATTRS` voi tarpeen mukaan lisätä pluginit ja attribuutit, joille makrotuksen haluaa sallia. En tiedä onko tuo varsinaisesti tarpeen, mutta ajattelin että olisi valmiiksi mekanismi millä voi rajata ongelmallisia makrotuksia pois mikäli sellaisia ilmenee.

Käsittääkseni jotkin jinja-filtterit toimivat jo nykyisin myös lohkon attribuuteissa, mutta tarvitseeko niitä ottaa tässä huomioon, kun ne käsitellään joka tapauksessa myöhemmin?

Reply:
> Olisiko kuinka hidasta tehdä niin, että kutsutaan tuo `expand_macros` jokaisen attribuutin kohdalla erikseen?

Niin joo ja tähän siis tarkennuksena, että tuo `expand_macros` ei mielestäni tee attribuutille käytännössä mitään jos attribuutin arvo ei tule makrolta. Toki se hyöty siinä voisi olla, että virheelliset makrokaavat jäisivät ehkä kiinni tuossa, mutta tuollaisia tuntemattomia attribuutteja ei kaiketi ole järkeä käsitellä?

Reply:
Lähinnä pointtini oli, että jos kirjoitan nyt

```
# {question=", b: 1,  "}
```

niin nyt tuo olisi sama kuin

```
# {question="" b="1"}
```

mikä ei mielestäni ole ihan toivottua. Sen takia kysyinkin, olisiko perää tehdä tuo makrojen laajennos jokaiselle attribuutin arvolle erikseen, jotta edellä mainittua ei voi tehdä. Asian voi esitellä myös toisin sanoin: mitä jos joskus käsiteltäväksi (eli makrolla avattavaksi) attribuutiksi lisätään sellainen, joka sallii arvokseen mielivaltaisen merkkijonon, ja joku laittaa sitten sen arvoksi `"Kissa, koira: hieno"`.

Reply:
> Toki se hyöty siinä voisi olla, että virheelliset makrokaavat jäisivät ehkä kiinni tuossa, mutta tuollaisia tuntemattomia attribuutteja ei kaiketi ole järkeä käsitellä?

Niin minäkin ajattelin, ja tosiaankin se, että ehkä makroavauksella ei haluaisi vahingossa lisätä uusia attribuutteja plugniin ihan tuosta vaan. Ei ainkanaan tällaisella "huijauksella" mielestäni.

Reply:
Ja siis tarkennan, että tosiaankin mielivaltaisen attribuutin saa lisättyä, mutta esittämäni esimerkki oli siitä, että sen nyt voi tehdä tuolloin "rikkinäisesti".

Reply:
Okei, joo, tuo selvensi ja samaa mieltä. Tuo pitää vielä siis tarkistaa, auttaako tuollaiseen huijausattribuuttiin tuo `expand_macros` pelkästään, vai tarvitseeko tehdä muutakin.

Reply:
Niinjoo

> Käsittääkseni jotkin jinja-filtterit toimivat jo nykyisin myös lohkon attribuuteissa, mutta tarvitseeko niitä ottaa tässä huomioon, kun ne käsitellään joka tapauksessa myöhemmin?

Joo, varmaan alkuun hyvä, että makrot ylipäätään toimivat. Filtterituki kyllä saadaan lisättyä sitten jos on tarve.

Reply:
Tämä nyt päivitetty tuohon ehdotettuun muotoon, löytyy myös <https://timdevs01-4.it.jyu.fi/lecture/users/test-user-1/test-qst>. Pikaisella testauksella 15, 20 tai 50 kpl satunnaistettua makrotettua QST-pluginia ei näyttänyt hidastavan (silmämääräisesti, en vielä mitannut tuota). Saattoi jopa nopeuttaa, koska tuossa jäi nyt pois aika paljon merkkijonojen käpistelyä.

Reply:
Jes, testasin samalla ohj1 monisteella enkä siitäkään huomannut muutoksia. Puolestani siis näyttää järkevältä